File: vivo_image_scraper.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import pandas as pd
import os
import time
from bs4 import BeautifulSoup
import requests

class VivoImageScraper:
    def __init__(self, excel_path):
        self.excel_path = excel_path
        self.output_dir = "product_images"
        os.makedirs(self.output_dir, exist_ok=True)
        
        # 初始化 Selenium WebDriver
        self.driver = webdriver.Chrome()
        print("\n初始化爬虫:")
        print(f"- Excel文件路径: {self.excel_path}")
        print(f"- 输出目录: {self.output_dir}")

    def scrape_product_images(self, product_name):
        try:
            # 访问搜索页面
            print("\n访问搜索页面...")
            self.driver.get("https://search.zol.com.cn/s/")
            
            # 等待搜索框和按钮加载
            wait = WebDriverWait(self.driver, 10)
            search_input = wait.until(
                EC.presence_of_element_located((By.ID, "keyword"))
            )
            search_button = wait.until(
                EC.presence_of_element_located((By.CLASS_NAME, "search-btn"))
            )
            
            # 输入搜索关键词
            print(f"输入搜索关键词: {product_name}")
            search_input.clear()
            search_input.send_keys(product_name)
            
            # 点击搜索按钮
            print("点击搜索按钮...")
            search_button.click()
            # 打断点
            # 等待搜索结果加载
            time.sleep(2)
            
            # 获取搜索结果页面内容
            search_doc = BeautifulSoup(self.driver.page_source, 'html.parser')
            # 查找产品标签
            product_tab = search_doc.select_one('.search-tab-nav.clearfix a[href*="detail.zol.com.cn"]')
            if product_tab:
                print("找到产品标签")
                # 如果不是当前选中的标签,需要点击切换
                if not 'current' in product_tab.get('class', []):
                    print("切换到产品标签...")
                    product_tab_element = self.driver.find_element(By.CSS_SELECTOR, '.search-tab-nav.clearfix a[href*="detail.zol.com.cn"]')
                    product_tab_element.click()
                    time.sleep(2)
                    # 重新获取页面内容
                    search_doc = BeautifulSoup(self.driver.page_source, 'html.parser')


            else:
                print("未找到产品标签")


            # 查找第一个产品链接
            product_link = search_doc.select_one('.pro-intro h3 a')
            if product_link:
                print("找到产品链接")
                # 获取href属性
                product_url = product_link.get('href')
                if product_url:
                    # 确保URL是完整的
                    if not product_url.startswith('http'):
                        product_url = 'https://detail.zol.com.cn' + product_url
                    print(f"产品URL: {product_url}")
                    
                    # 点击链接访问产品页面
                    try:
                        product_link_element = self.driver.find_element(By.CSS_SELECTOR, '.pro-intro h3 a')
                        product_link_element.click()
                        time.sleep(2)
                        # 获取当前页面内容
                        product_doc = BeautifulSoup(self.driver.page_source, 'html.parser')
                        print("产品页面内容:")
                        # 打印断点调试信息
                    
                        print(product_doc.prettify())
                        # 查找参数链接
                        param_link = search_doc.select_one('.nav__list a[href*="/param.shtml"]')
                        if param_link:
                            print("找到参数链接")
                            param_url = param_link.get('href')
                            if param_url:
                                # 确保URL是完整的
                                if not param_url.startswith('http'):
                                    param_url = 'https://detail.zol.com.cn' + param_url
                                print(f"参数页面URL: {param_url}")
                                
                                # 点击链接访问参数页面
                                try:
                                    # 使用更精确的选择器定位参数链接
                                    param_link_element = self.driver.find_element(By.CSS_SELECTOR, '.nav__list.clearfix li a[href*="/param.shtml"]')
                                    param_link_element.click()
                                    time.sleep(2)
                                    
                                    # 获取参数页面内容
                                    param_doc = BeautifulSoup(self.driver.page_source, 'html.parser')
                                    print("参数页面内容:")
                                    print(param_doc.prettify())
                                except Exception as e:
                                    print(f"点击参数链接失败: {str(e)}")
                                    # 如果点击失败,直接访问URL
                                    self.driver.get(param_url)
                                    time.sleep(2)
                                    
                                    # 获取参数页面内容
                                    param_doc = BeautifulSoup(self.driver.page_source, 'html.parser')
                                    print("参数页面内容:")
                                    print(param_doc.prettify())
                            else:
                                print("参数链接href属性为空")
                        else:
                            print("未找到参数链接")
                    except Exception as e:
                        print(f"点击产品链接失败: {str(e)}")
                        # 如果点击失败,直接访问URL
                        self.driver.get(product_url)
                        time.sleep(2)
                else:
                    print("产品链接href属性为空")
            else:
                print("未找到产品链接")
            # 创建产品目录

            # 获取当前页面内容

            # 查找参数链接
            param_link = search_doc.select_one('.nav__list a[href*="/param.shtml"]')
            if param_link:
                print("找到参数链接")
                param_url = param_link.get('href')
                if param_url:
                    # 确保URL是完整的
                    if not param_url.startswith('http'):
                        param_url = 'https://detail.zol.com.cn' + param_url
                    print(f"参数页面URL: {param_url}")
                    
                    # 点击链接访问参数页面
                    try:
                        # 使用更精确的选择器定位参数链接
                        param_link_element = self.driver.find_element(By.CSS_SELECTOR, '.nav__list.clearfix li a[href*="/param.shtml"]')
                        param_link_element.click()
                        time.sleep(2)
                        
                        # 获取参数页面内容
                        param_doc = BeautifulSoup(self.driver.page_source, 'html.parser')
                        print("参数页面内容:")
                        print(param_doc.prettify())
                    except Exception as e:
                        print(f"点击参数链接失败: {str(e)}")
                        # 如果点击失败,直接访问URL
                        self.driver.get(param_url)
                        time.sleep(2)
                        
                        # 获取参数页面内容
                        param_doc = BeautifulSoup(self.driver.page_source, 'html.parser')
                        print("参数页面内容:")
                        print(param_doc.prettify())
                else:
                    print("参数链接href属性为空")
            else:
                print("未找到参数链接")

            # 创建产品目录
            product_dir = os.path.join('images', product_name)
            os.makedirs(product_dir, exist_ok=True)
            print(f"创建目录: {product_dir}")
            
            # 查找所有图片元素
            image_elements = product_doc.select('.big-pic-box img')
            if not image_elements:
                print("未找到图片元素")
                return
                
            print(f"找到 {len(image_elements)} 张图片")
            
            # 下载每张图片
            for i, img in enumerate(image_elements, 1):
                img_url = img.get('src')
                if not img_url:
                    print(f"图片 {i} 没有src属性")
                    continue
                    
                # 确保URL是完整的
                if not img_url.startswith('http'):
                    img_url = 'https:' + img_url
                    
                # 构建保存路径
                img_name = f"{product_name}_{i}.jpg"
                img_path = os.path.join(product_dir, img_name)
                
                try:
                    # 下载图片
                    response = requests.get(img_url, timeout=10)
                    if response.status_code == 200:
                        with open(img_path, 'wb') as f:
                            f.write(response.content)
                        print(f"已保存图片: {img_name}")
                    else:
                        print(f"下载图片 {img_name} 失败: HTTP {response.status_code}")
                except Exception as e:
                    print(f"下载图片 {img_name} 时出错: {str(e)}")
        except Exception as e:
            print(f"处理产品 {product_name} 时发生错误:")
            print(f"- 错误类型: {type(e).__name__}")
            print(f"- 错误信息: {str(e)}")
            import traceback
            print("- 错误堆栈:")
            print('\n'.join(traceback.format_stack()[:6]))

    def scrape_images(self):
        try:
            # 读取Excel文件
            df = pd.read_excel(self.excel_path)
            
            # 打印列名，帮助调试
            print("\nExcel文件包含以下列:")
            print(df.columns.tolist())
            
            # 使用 "名称" 列而不是 "产品名称" 列
            if '名称' not in df.columns:
                print("\n错误: Excel文件中没有找到'名称'列")
                print("请确保Excel文件中包含'名称'列，或修改以下列名之一:")
                for col in df.columns:
                    print(f"- {col}")
                return
            
            # 遍历每个产品
            for index, row in df.iterrows():
                product_name = row['名称']  # 使用 "名称" 而不是 "产品名称"
                print(f"\n处理第 {index + 1} 个产品: {product_name}")
                self.scrape_product_images(product_name)
                
        except Exception as e:
            print(f"\n读取Excel文件时发生错误: {str(e)}")
            print("请确保:")
            print("1. Excel文件存在且可访问")
            print("2. 文件格式正确（.xlsx或.xls）")
            print("3. 文件未被其他程序占用")
        finally:
            self.close()

    def close(self):
        if hasattr(self, 'driver'):
            self.driver.quit()

if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print("请提供Excel文件路径，例如: python vivo_image_scraper.py products.xlsx")
    else:
        scraper = VivoImageScraper(sys.argv[1])
        try:
            scraper.scrape_images()
        except KeyboardInterrupt:
            print("\n程序被用户中断")
        finally:
            scraper.close() 

File: vivo_image_scraper.js
const puppeteer = require('puppeteer');
const xlsx = require('xlsx');
const fs = require('fs');
const path = require('path');
const axios = require('axios');

class VivoImageScraper {
  constructor(excelPath) {
    this.excelPath = excelPath;
    this.outputDir = "product_images";
    fs.mkdirSync(this.outputDir, { recursive: true });

    console.log("\n初始化爬虫:");
    console.log(`- Excel文件路径: ${this.excelPath}`);
    console.log(`- 输出目录: ${this.outputDir}`);
  }

  async delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  async scrapeProductImages(productName, browser) {
    try {
      // 创建新页面
      const page = await browser.newPage();

      // 访问搜索页面
      console.log("\n访问搜索页面...");
      await page.goto("https://search.zol.com.cn/s/");

      // 等待搜索框和按钮加载
      await page.waitForSelector("#keyword");
      await page.waitForSelector(".search-btn");

      // 输入搜索关键词
      console.log(`输入搜索关键词: ${productName}`);
      await page.type("#keyword", productName);

      // 点击搜索按钮
      console.log("点击搜索按钮...");
      await page.click(".search-btn");

      // 等待搜索结果加载
      await this.delay(2000);

      // 查找产品标签
      const productTab = await page.$('.search-tab-nav.clearfix a[href*="detail.zol.com.cn"]');
      if (productTab) {
        console.log("找到产品标签");
        // 检查是否需要点击切换
        const isCurrentTab = await productTab.evaluate(el => el.classList.contains('current'));
        if (!isCurrentTab) {
          console.log("切换到产品标签...");
          await productTab.click();
          await this.delay(2000);
        }
      } else {
        console.log("未找到产品标签");
      }

      // 查找第一个产品链接
      const productLink = await page.$('.pro-intro h3 a');
      if (productLink) {
        console.log("找到产品链接");
        const productUrl = await productLink.evaluate(el => el.href);

        if (productUrl) {
          console.log(`产品URL: ${productUrl}`);

          // 获取当前页面句柄
          const currentPageTarget = page.target();

          // 点击链接访问产品页面
          await productLink.click();
          await this.delay(2000);

          // 等待新页面打开
          const newTarget = await browser.waitForTarget(target => target.opener() === currentPageTarget);
          const newPage = await newTarget.page();

          // 切换到新页面
          await newPage.bringToFront();

          // 获取新页面内容
          const content = await newPage.content();
          console.log('页面内容:', content);
          // 将页面内容写入文件
          const contentPath = path.join('content.txt');
          fs.writeFileSync(contentPath, content, 'utf8');
          console.log(`页面内容已保存到: ${contentPath}`);
          // 查找并点击参数链接
          const paramLink = await newPage.$('a[href*="/param.shtml"]');
          if (paramLink) {
            console.log("找到参数链接");
            await paramLink.click();
            await this.delay(2000);

            // 等待新页面加载
            const paramPageTarget = await browser.waitForTarget(target => target.url().includes('/param.shtml'));
            const paramPage = await paramPageTarget.page();
            await paramPage.bringToFront();

            // 获取参数页面内容
            const paramContent = await paramPage.content();
            console.log('参数页面内容已获取');

            // 获取参数表格内容
            const paramTable = await paramPage.$('.detailed-parameters table');
            if (paramTable) {
                const tableData = await paramPage.evaluate(table => {
                    const rows = Array.from(table.querySelectorAll('tr'));
                    return rows.map(row => {
                        const cells = Array.from(row.querySelectorAll('th, td'));
                        return cells.map(cell => cell.textContent.trim()).join('\t');
                    }).join('\n');
                }, paramTable);
                
                // 获取所有Excel文件
                const excelFiles = fs.readdirSync('.').filter(file => file.endsWith('.xlsx'));
                
                // 为每个Excel文件创建对应的目录并保存参数
                for (const excelFile of excelFiles) {
                    // 获取Excel文件名（不含扩展名）
                    const excelFileName = path.basename(excelFile, '.xlsx');
                    
                    // 创建Excel文件对应的目录
                    const excelDir = path.join('params', excelFileName);
                    fs.mkdirSync(excelDir, { recursive: true });
                    
                    // 在Excel目录下创建产品目录
                    const productDir = path.join(excelDir, productName.replace(/[/\\]/g, '_'));
                    fs.mkdirSync(productDir, { recursive: true });
                    
                    // 保存参数表格内容到对应产品文件夹
                    const paramContentPath = path.join(productDir, 'params.txt');
                    fs.writeFileSync(paramContentPath, tableData, 'utf8');
                    console.log(`参数表格内容已保存到: ${paramContentPath}`);
                }
            } else {
                console.log("未找到参数表格");
            }
            await paramPage.close();
          } else {
            console.log("未找到参数链接");
          }




          // 创建产品目录
          const productDir = path.join('images', productName.replace(/[/\\]/g, '_'));
          fs.mkdirSync(productDir, { recursive: true });
          console.log(`创建目录: ${productDir}`);

          // 查找所有图片元素
          const imageElements = await page.$$('.big-pic-box img');
          if (!imageElements.length) {
            console.log("未找到图片元素");
            return;
          }

          console.log(`找到 ${imageElements.length} 张图片`);

          // 下载每张图片
          for (let i = 0; i < imageElements.length; i++) {
            const imgUrl = await imageElements[i].evaluate(el => el.src);
            if (!imgUrl) {
              console.log(`图片 ${i + 1} 没有src属性`);
              continue;
            }

            // 确保URL是完整的
            const fullImgUrl = imgUrl.startsWith('http') ? imgUrl : `https:${imgUrl}`;

            // 构建保存路径
            const imgName = `${productName.replace(/[/\\]/g, '_')}_${i + 1}.jpg`;
            const imgPath = path.join(productDir, imgName);

            try {
              // 下载图片
              const response = await axios({
                method: 'get',
                url: fullImgUrl,
                responseType: 'stream',
                timeout: 10000
              });

              if (response.status === 200) {
                const writer = fs.createWriteStream(imgPath);
                response.data.pipe(writer);

                await new Promise((resolve, reject) => {
                  writer.on('finish', resolve);
                  writer.on('error', reject);
                });

                console.log(`已保存图片: ${imgName}`);
              } else {
                console.log(`下载图片 ${imgName} 失败: HTTP ${response.status}`);
              }
            } catch (error) {
              console.log(`下载图片 ${imgName} 时出错: ${error.message}`);
            }
          }
        } else {
          console.log("产品链接href属性为空");
        }
      } else {
        console.log("未找到产品链接");
      }

      await page.close();
    } catch (error) {
      console.log(`处理产品 ${productName} 时发生错误:`);
      console.log(`- 错误类型: ${error.constructor.name}`);
      console.log(`- 错误信息: ${error.message}`);
      console.log("- 错误堆栈:");
      console.log(error.stack.split('\n').slice(0, 6).join('\n'));
    }
  }

  async scrapeImages() {
    let browser;
    try {
      // 读取Excel文件
      const workbook = xlsx.readFile(this.excelPath);
      const worksheet = workbook.Sheets[workbook.SheetNames[0]];
      const data = xlsx.utils.sheet_to_json(worksheet);

      // 打印列名，帮助调试
      console.log("\nExcel文件包含以下列:");
      console.log(Object.keys(data[0]));

      // 检查是否存在"名称"列
      if (!data[0].hasOwnProperty('名称')) {
        console.log("\n错误: Excel文件中没有找到'名称'列");
        console.log("请确保Excel文件中包含'名称'列，或修改以下列名之一:");
        Object.keys(data[0]).forEach(col => {
          console.log(`- ${col}`);
        });
        return;
      }

      // 启动浏览器
      browser = await puppeteer.launch({
        headless: false,
        defaultViewport: null
      });

      // 遍历每个产品
      for (let i = 0; i < data.length; i++) {
        const productName = data[i]['名称'];
        console.log(`\n处理第 ${i + 1} 个产品: ${productName}`);
        await this.scrapeProductImages(productName, browser);
      }
    } catch (error) {
      console.log(`\n读取Excel文件时发生错误: ${error.message}`);
      console.log("请确保:");
      console.log("1. Excel文件存在且可访问");
      console.log("2. 文件格式正确（.xlsx或.xls）");
      console.log("3. 文件未被其他程序占用");
    } finally {
      if (browser) {
        await browser.close();
      }
    }
  }
}

// 运行爬虫
if (process.argv.length < 3) {
  console.log("请提供Excel文件路径，例如: node vivo_image_scraper.js products.xlsx");
} else {
  const scraper = new VivoImageScraper(process.argv[2]);
  scraper.scrapeImages().catch(error => {
    console.log("\n程序执行出错:", error);
  });
} 
